File: python/core/engine.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
SQL query execution engine with interactive REPL
"""

import sqlite3
from typing import List, Dict, Any, Optional
from .schema import SchemaInference


class SQLEngine:
    """
    SQL query execution engine with interactive REPL.
    Extracted from query_json_with_sql() in jsonsql.py
    """

    def __init__(self):
        self.conn: Optional[sqlite3.Connection] = None
        self.cursor: Optional[sqlite3.Cursor] = None

    def load_data(self, data: List[Dict[str, Any]], table_name: str = "data") -> None:
        """
        Load data into in-memory SQLite database.

        Args:
            data: List of dictionaries to load
            table_name: Name for the table
        """
        self.conn = sqlite3.connect(':memory:')
        self.cursor = self.conn.cursor()

        SchemaInference.create_table_from_data(
            self.cursor,
            table_name,
            data
        )

        print(f"已加载 {len(data)} 条记录到表 '{table_name}' 中")

    def execute_query(self, sql_query: str) -> Optional[List[tuple]]:
        """
        Execute a SQL query.

        Args:
            sql_query: SQL query string

        Returns:
            Query results or None for non-SELECT queries
        """
        if not self.cursor:
            raise RuntimeError("No data loaded. Call load_data() first.")

        self.cursor.execute(sql_query)

        if sql_query.strip().upper().startswith('SELECT'):
            return self.cursor.fetchall()
        else:
            self.conn.commit()
            return None

    def get_column_names(self) -> Optional[List[str]]:
        """Get column names from last query"""
        if self.cursor and self.cursor.description:
            return [desc[0] for desc in self.cursor.description]
        return None

    def run_repl(self, table_name: str = "data") -> None:
        """
        Run interactive REPL for SQL queries.

        This is extracted from the main loop in query_json_with_sql()

        Args:
            table_name: Default table name for queries
        """
        print("请输入SQL查询语句，输入 'exit' 或 'quit' 退出程序\n")

        while True:
            try:
                sql_query = input("SQL> ").strip()

                if sql_query.lower() in ['exit', 'quit', '']:
                    break

                results = self.execute_query(sql_query)

                if sql_query.strip().upper().startswith('SELECT'):
                    column_names = self.get_column_names()

                    if column_names:
                        print(f"\n列名: {column_names}")

                    if results:
                        print(f"找到 {len(results)} 条记录:")
                        for i, row in enumerate(results, 1):
                            print(f"记录 {i}: {dict(zip(column_names, row))}")
                    else:
                        print("未找到匹配的记录")
                else:
                    print(f"执行完成，影响了 {self.cursor.rowcount} 行")

            except sqlite3.Error as e:
                print(f"SQL错误: {e}")
            except KeyboardInterrupt:
                print("\n程序被中断")
                break
            except Exception as e:
                print(f"错误: {e}")

    def close(self) -> None:
        """Close database connection"""
        if self.conn:
            self.conn.close()

File: nodejs/src/core/engine.ts
import initSqlJs, { Database, SqlJsStatic } from 'sql.js';
import * as readline from 'readline';
import { ParserData } from '../parsers/base';
import { SchemaInference } from './schema';

export class SQLEngine {
  private db: Database | null = null;
  private SQL: SqlJsStatic | null = null;

  async init(): Promise<void> {
    this.SQL = await initSqlJs();
    this.db = new this.SQL.Database();
  }

  async loadData(data: ParserData[], tableName: string = 'data'): Promise<void> {
    if (!this.SQL) {
      await this.init();
    }
    if (!this.db) {
      throw new Error('Database not initialized');
    }

    await SchemaInference.createTableFromData(this.db, tableName, data);
    console.log(`Loaded ${data.length} records into table '${tableName}'`);
  }

  executeQuery(sqlQuery: string): any[] | null {
    if (!this.db) {
      throw new Error('No data loaded. Call loadData() first.');
    }

    const trimmedQuery = sqlQuery.trim().toUpperCase();
    const isSelect = trimmedQuery.startsWith('SELECT');

    if (isSelect) {
      const stmt = this.db.exec(sqlQuery);
      if (stmt.length === 0) {
        return [];
      }

      const columns = stmt[0].columns;
      const values = stmt[0].values;

      return values.map((row: any[]) => {
        const obj: any = {};
        columns.forEach((col: string, index: number) => {
          obj[col] = row[index];
        });
        return obj;
      });
    } else {
      this.db.run(sqlQuery);
      return null;
    }
  }

  getColumnNames(lastQuery: string): string[] | null {
    if (!this.db) {
      return null;
    }

    try {
      const stmt = this.db.exec(lastQuery);
      if (stmt.length === 0) {
        return null;
      }
      return stmt[0].columns;
    } catch {
      return null;
    }
  }

  async runRepl(tableName: string = 'data'): Promise<void> {
    console.log('\nEnter SQL queries, type \'exit\' or \'quit\' to exit\n');

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    const askQuestion = (query: string): Promise<string> => {
      return new Promise((resolve) => {
        rl.question(query, (answer: string) => {
          resolve(answer);
        });
      });
    };

    while (true) {
      try {
        const sqlQuery = await askQuestion('SQL> ');
        const trimmedQuery = sqlQuery.trim();

        if (trimmedQuery.toLowerCase() === 'exit' || trimmedQuery.toLowerCase() === 'quit' || trimmedQuery === '') {
          break;
        }

        const results = this.executeQuery(sqlQuery);

        if (trimmedQuery.toUpperCase().startsWith('SELECT')) {
          const columnNames = this.getColumnNames(sqlQuery);

          if (columnNames) {
            console.log(`\nColumns: [${columnNames.join(', ')}]`);
          }

          if (results && results.length > 0) {
            console.log(`Found ${results.length} records:`);
            for (let i = 0; i < results.length; i++) {
              console.log(`Record ${i + 1}: ${JSON.stringify(results[i])}`);
            }
          } else {
            console.log('No matching records found');
          }
        } else {
          console.log('Execution completed');
        }
      } catch (error) {
        console.error(`SQL Error: ${error instanceof Error ? error.message : String(error)}`);
      }
    }

    rl.close();
  }

  close(): void {
    if (this.db) {
      this.db.close();
      this.db = null;
    }
  }
}
